File: src/test/java/org/blitzortung/android/alert/handler/AlertSectorHandlerTest.java
package org.blitzortung.android.alert.handler;

import android.location.Location;
import com.google.common.collect.Lists;
import org.blitzortung.android.alert.AlertParameters;
import org.blitzortung.android.alert.object.AlertSector;
import org.blitzortung.android.alert.object.AlertSectorRange;
import org.blitzortung.android.data.beans.Strike;
import org.blitzortung.android.util.MeasurementSystem;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyFloat;
import static org.mockito.Mockito.*;

@RunWith(RobolectricTestRunner.class)
public class AlertSectorHandlerTest {
    
    @Mock
    private Strike strike;
    
    @Mock
    private Location location;

    @Mock
    private Location strikeLocation;

    private long now;

    private long thresholdTime;

    @Mock
    private AlertSector alertSector;
    
    @Mock
    private AlertSectorRange alertSectorRange1;

    @Mock
    private AlertSectorRange alertSectorRange2;
    
    @Mock
    private AlertParameters alertParameters;
    
    private final MeasurementSystem measurementSystem = MeasurementSystem.METRIC;
    
    private AlertSectorHandler alertSectorHandler;
    private long beforeThresholdTime;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        
        now = System.currentTimeMillis();
        thresholdTime = now - 10 * 60 * 1000;
        beforeThresholdTime = thresholdTime - 1;
        
        alertSectorHandler = new AlertSectorHandler(alertParameters);
        alertSectorHandler.setCheckStrikeParameters(location, thresholdTime);
        
        when(alertSector.getRanges()).thenReturn(Lists.newArrayList(alertSectorRange1, alertSectorRange2));
        when(alertParameters.getMeasurementSystem()).thenReturn(measurementSystem);
        when(alertSectorRange1.getRangeMaximum()).thenReturn(2.5f);
        when(alertSectorRange2.getRangeMaximum()).thenReturn(5f);
    }
    
    @Test
    public void testCheckWithNullAsSector()
    {
        alertSectorHandler.checkStrike(null, strike);
        
        verify(location, times(0)).distanceTo(any(Location.class));
        verify(alertParameters, times(0)).getMeasurementSystem();
    }
    
    @Test
    public void testCheckWithinThresholdTimeAndRange1()
    {
        when(strike.getTimestamp()).thenReturn(thresholdTime);
        when(strike.getLocation(any(Location.class))).thenReturn(strikeLocation);
        when(location.distanceTo(strikeLocation)).thenReturn(2500f);
        
        alertSectorHandler.checkStrike(alertSector, strike);
        
        verify(alertSector, times(1)).updateClosestStrikeDistance(2.5f);
        verify(alertSectorRange1, times(1)).getRangeMaximum();
        verify(alertSectorRange1, times(1)).addStrike(strike);
        verify(alertSectorRange2, times(0)).getRangeMaximum();
    }

    @Test
    public void testCheckWithinThresholdTimeAndOutOfAllRanges()
    {
        when(strike.getTimestamp()).thenReturn(thresholdTime);
        when(strike.getLocation(any(Location.class))).thenReturn(strikeLocation);
        when(location.distanceTo(strikeLocation)).thenReturn(5000.1f);

        alertSectorHandler.checkStrike(alertSector, strike);

        verify(alertSector, times(0)).updateClosestStrikeDistance(anyFloat());
        verify(alertSectorRange1, times(1)).getRangeMaximum();
        verify(alertSectorRange1, times(0)).addStrike(any(Strike.class));
        verify(alertSectorRange2, times(1)).getRangeMaximum();
        verify(alertSectorRange2, times(0)).addStrike(any(Strike.class));
    }

    @Test
    public void testCheckOutOfThresholdTimeAndWithinRange2()
    {
        when(strike.getTimestamp()).thenReturn(beforeThresholdTime);
        when(strike.getLocation(any(Location.class))).thenReturn(strikeLocation);
        when(location.distanceTo(strikeLocation)).thenReturn(2500.1f);

        alertSectorHandler.checkStrike(alertSector, strike);

        verify(alertSector, times(0)).updateClosestStrikeDistance(anyFloat());
        verify(alertSectorRange1, times(1)).getRangeMaximum();
        verify(alertSectorRange1, times(0)).addStrike(any(Strike.class));
        verify(alertSectorRange2, times(1)).getRangeMaximum();
        verify(alertSectorRange2, times(1)).addStrike(strike);
    }

    @Test
    public void testCheckOutOfThresholdTimeAndAllRanges()
    {
        when(strike.getTimestamp()).thenReturn(beforeThresholdTime);
        when(strike.getLocation(any(Location.class))).thenReturn(strikeLocation);
        when(location.distanceTo(strikeLocation)).thenReturn(5000.1f);

        alertSectorHandler.checkStrike(alertSector, strike);

        verify(alertSector, times(0)).updateClosestStrikeDistance(anyFloat());
        verify(alertSectorRange1, times(1)).getRangeMaximum();
        verify(alertSectorRange1, times(0)).addStrike(any(Strike.class));
        verify(alertSectorRange2, times(1)).getRangeMaximum();
        verify(alertSectorRange2, times(0)).addStrike(any(Strike.class));
    }
}

File: app/src/test/java/org/blitzortung/android/alert/handler/AlertSectorHandlerTest.kt
package org.blitzortung.android.alert.handler

import android.location.Location
import org.blitzortung.android.alert.AlertParameters
import org.blitzortung.android.alert.data.AlertContext
import org.blitzortung.android.alert.data.AlertSector
import org.blitzortung.android.alert.data.AlertSectorRange
import org.blitzortung.android.data.beans.Strike
import org.blitzortung.android.util.MeasurementSystem
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Matchers.any
import org.mockito.Matchers.anyFloat
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner
import org.assertj.core.api.Assertions.assertThat
import org.blitzortung.android.data.beans.DefaultStrike

@RunWith(RobolectricTestRunner::class)
class AlertSectorHandlerTest {

    private val measurementSystem = MeasurementSystem.METRIC

    private lateinit var strike: DefaultStrike

    @Mock
    private lateinit var location: Location

    private lateinit var strikeLocation: Location

    private var now: Long = 0

    private var thresholdTime: Long = 0

    private lateinit var alertSectorRange1: AlertSectorRange

    private lateinit var alertSectorRange2: AlertSectorRange

    private lateinit var alertParameters: AlertParameters

    private lateinit var alertSectorHandler: AlertSectorHandler

    private lateinit var alertSector: AlertSector

    private lateinit var alertContext: AlertContext

    private lateinit var sectors: List<AlertSector>

    private var beforeThresholdTime: Long = 0

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        strike = DefaultStrike(timestamp = now - 200, longitude = 12.0f, latitude = 41.0f, altitude = 0, amplitude = 20.0f)

        strikeLocation = Location("")
        strikeLocation.longitude = strike.longitude.toDouble()
        strikeLocation.latitude = strike.latitude.toDouble()

        alertSectorRange1 = AlertSectorRange(rangeMinimum = 0.0f, rangeMaximum = 2.5f)
        alertSectorRange2 = AlertSectorRange(rangeMinimum = 2.5f, rangeMaximum = 5.0f)

        alertSector = AlertSector(
                label = "SO",
                minimumSectorBearing = 12.0f,
                maximumSectorBearing = 23.5f,
                ranges = listOf(alertSectorRange1, alertSectorRange2)
        )

        sectors = listOf(alertSector)

        alertParameters = AlertParameters(
                alarmInterval = 10,
                rangeSteps = arrayOf(2.5f, 5f),
                sectorLabels = arrayOf("SO"),
                measurementSystem = MeasurementSystem.METRIC
        )

        alertContext = AlertContext(location, alertParameters, sectors)

        now = System.currentTimeMillis()
        thresholdTime = now - 10 * 60 * 1000
        beforeThresholdTime = thresholdTime - 1

        alertSectorHandler = AlertSectorHandler()
        alertSectorHandler.setCheckStrikeParameters(location, thresholdTime)
    }

    @Test
    fun testCheckWithinThresholdTimeAndRange1() {
        strike = strike.copy(timestamp = thresholdTime)

        `when`(location.distanceTo(strikeLocation)).thenReturn(2500f)

        alertSectorHandler.checkStrike(alertSector, strike, measurementSystem)

        assertThat(alertSector.closestStrikeDistance).isEqualTo(2.5f)
    }

    @Test
    fun testCheckWithinThresholdTimeAndOutOfAllRanges() {
        strike = strike.copy(timestamp = thresholdTime)
        `when`(location.distanceTo(strikeLocation)).thenReturn(5000.1f)

        alertSectorHandler.checkStrike(alertSector, strike, measurementSystem)

        assertThat(alertSector.closestStrikeDistance).isEqualTo(Float.POSITIVE_INFINITY)
    }

    @Test
    fun testCheckOutOfThresholdTimeAndWithinRange2() {
        strike = strike.copy(timestamp = beforeThresholdTime)
        `when`(location.distanceTo(strikeLocation)).thenReturn(2500.1f)

        alertSectorHandler.checkStrike(alertSector, strike, measurementSystem)

        assertThat(alertSector.closestStrikeDistance).isEqualTo(Float.POSITIVE_INFINITY)
    }

    @Test
    fun testCheckOutOfThresholdTimeAndAllRanges() {
        strike = strike.copy(timestamp = beforeThresholdTime)
        `when`(location.distanceTo(strikeLocation)).thenReturn(5000.1f)

        alertSectorHandler.checkStrike(alertSector, strike, measurementSystem)

        assertThat(alertSector.closestStrikeDistance).isEqualTo(Float.POSITIVE_INFINITY)
    }
}
